File: web/src/erp/utils/businessProgress.mjs
const integer = (value) => Number.isSafeInteger(value) && value >= 0
const decimal = (value) =>
  value === null || (typeof value === 'string' && /^\d+(\.\d+)?$/u.test(value))
const counters = [
  'product_count',
  'engineering_total',
  'engineering_ready',
  'production_orders',
  'production_closed',
  'material_total',
  'material_pending',
  'in_progress_batches',
  'outsourced_batches',
  'waiting_batches',
  'rejected_batches',
  'planned_batches',
  'open_tasks',
  'blocked_tasks',
  'unassigned_tasks',
  'attention_task_id',
  'operation_count',
]

function invalid() {
  return Object.assign(new Error('进度数据暂不可用，请重试'), {
    isInvalidResponse: true,
  })
}
function validEnvelope(value) {
  return (
    value &&
    typeof value.snapshot_at === 'string' &&
    Number.isFinite(Date.parse(value.snapshot_at)) &&
    ['sales', 'production', 'tasks', 'wip'].every(
      (key) => typeof value.access?.[key] === 'boolean'
    )
  )
}
function validRow(row) {
  return (
    row &&
    integer(row.id) &&
    row.id > 0 &&
    ['orders', 'production'].includes(row.view) &&
    [
      'order_no',
      'status',
      'customer',
      'product',
      'sales_owner',
      'unit',
      'attention_task',
      'attention_reason',
      'attention_owner',
      'attention_role',
      'due_date',
      'updated_at',
      'current_operation',
    ].every((key) => typeof row[key] === 'string') &&
    integer(row.product_id) &&
    counters.every((key) => integer(row[key])) &&
    [
      'active',
      'overdue',
      'due_soon',
      'blocked',
      'unlinked',
      'delivery_known',
    ].every((key) => typeof row[key] === 'boolean') &&
    decimal(row.ordered_quantity) &&
    decimal(row.shipped_quantity) &&
    decimal(row.completed_quantity) &&
    (row.delivery_known || row.shipped_quantity === null)
  )
}
export function requireProgressBoard(value) {
  if (
    !validEnvelope(value) ||
    !Array.isArray(value.rows) ||
    !integer(value.total) ||
    !['total', 'overdue', 'due_soon', 'blocked', 'undated'].every((key) =>
      integer(value.counts?.[key])
    ) ||
    value.total > value.counts.total ||
    value.rows.length > value.total ||
    !value.rows.every(validRow) ||
    new Set(value.rows.map((row) => row.id)).size !== value.rows.length
  ) {
    throw invalid()
  }
  return value
}
export function requireProgressDetail(value) {
  if (
    !validEnvelope(value) ||
    !validRow(value.row) ||
    !value.sections ||
    !value.has_more
  ) {
    throw invalid()
  }
  for (const [key, rows] of Object.entries(value.sections)) {
    if (
      !['lines', 'production', 'materials', 'batches', 'tasks'].includes(key) ||
      !Array.isArray(rows) ||
      rows.length > 100 ||
      !rows.every(
        (row) =>
          row &&
          integer(row.id) &&
          row.id > 0 &&
          integer(row.parent_id) &&
          [
            'kind',
            'number',
            'label',
            'status',
            'quantity',
            'unit',
            'date',
            'owner',
            'role',
            'note',
          ].every((field) => typeof row[field] === 'string')
      )
    ) {
      throw invalid()
    }
  }
  return value
}

export function progressQueryFromURL(params) {
  const allowed = (key, values, fallback) =>
    values.includes(params.get(key)) ? params.get(key) : fallback
  const date = (key) =>
    /^\d{4}-\d{2}-\d{2}$/u.test(params.get(key) || '') ? params.get(key) : ''
  const page = Math.max(
    1,
    Math.min(50000, Number.parseInt(params.get('page'), 10) || 1)
  )
  return {
    view: allowed('view', ['orders', 'production'], ''),
    keyword: (params.get('q') || '').slice(0, 100),
    scope: allowed('scope', ['active', 'all', 'ended'], 'active'),
    risk: allowed(
      'risk',
      ['all', 'overdue', 'due_soon', 'blocked', 'undated', 'unlinked'],
      'all'
    ),
    owner: (params.get('owner') || '').slice(0, 100),
    date_from: date('from'),
    date_to: date('to'),
    limit: 20,
    offset: (page - 1) * 20,
  }
}

const statusLabels = {
  draft: '待提交',
  submitted: '待生效',
  active: '已生效',
  closed: '已关闭',
  canceled: '已取消',
  DRAFT: '待下达',
  RELEASED: '已下达',
  CLOSED: '已关闭',
  CANCELLED: '已取消',
  PREPARING: '资料准备',
  SAMPLING: '打样中',
  CONFIRMED: '已确认',
  PLANNED: '待开工',
  IN_PROGRESS: '生产中',
  OUTSOURCED: '委外中',
  WAITING_QUALITY: '待检',
  ACCEPTED: '已验收',
  REJECTED: '不合格',
  PENDING: '尚未领齐',
  ISSUED: '按计划领齐',
  ready: '待处理',
  blocked: '阻塞',
  done: '已完成',
  rejected: '已退回',
  withdrawn: '已撤回',
}
export const progressStatusLabel = (status) => statusLabels[status] || '待核对'
export const progressQuantity = (value) =>
  value === null || value === undefined || value === '' ? '—' : String(value)

export function progressDelivery(row) {
  if (row.view !== 'orders') {
    if (row.completed_quantity === null || row.ordered_quantity === null) {
      return {
        label: '计划数量',
        text:
          row.ordered_quantity === null
            ? '按明细查看'
            : `${row.ordered_quantity} ${row.unit}`,
      }
    }
    const planned = Number(row.ordered_quantity)
    const completed = Number(row.completed_quantity)
    return {
      label: '有效完工 / 计划',
      text: `${row.completed_quantity} / ${row.ordered_quantity} ${row.unit}`,
      percent:
        planned > 0 ? Math.min(100, (completed / planned) * 100) : undefined,
    }
  }
  if (!row.delivery_known) return { label: '出货进度', text: '关联待核对' }
  if (row.ordered_quantity === null || row.shipped_quantity === null) {
    return { label: '出货进度', text: '多单位 · 查看明细' }
  }
  const ordered = Number(row.ordered_quantity)
  const shipped = Number(row.shipped_quantity)
  if (
    !(ordered > 0) ||
    !Number.isFinite(ordered) ||
    !Number.isFinite(shipped)
  ) {
    return { label: '出货进度', text: '待核对' }
  }
  return {
    label: '已出货 / 订购',
    text: `${progressQuantity(row.shipped_quantity)} / ${progressQuantity(
      row.ordered_quantity
    )} ${row.unit}`,
    percent: Math.min(100, (shipped / ordered) * 100),
    complete: shipped >= ordered,
  }
}

export function progressSourcePath(record, view = 'orders') {
  if (record.kind === 'task') {
    return `/erp/task-board?q=${encodeURIComponent(record.number)}&status=all`
  }
  if (record.kind === 'sales_line') {
    return `/erp/sales/project-orders/sales-orders?sales_order_id=${record.parent_id}`
  }
  if (record.kind === 'batch') {
    return `/erp/production/orders?production_order_id=${
      record.parent_id
    }&wip_batch_id=${record.id}`
  }
  if (record.kind === 'production_line') {
    return `/erp/production/orders?production_order_id=${
      record.parent_id
    }&production_order_item_id=${record.id}`
  }
  if (record.kind === 'material') {
    return `/erp/production/orders?production_order_id=${record.parent_id}`
  }
  if (record.kind === 'production' || view === 'production') {
    return `/erp/production/orders?production_order_id=${record.id}`
  }
  return `/erp/sales/project-orders/sales-orders?sales_order_id=${record.id}`
}

File: server/internal/biz/business_progress.go
package biz

import (
	"context"
	"strings"
	"time"
)

// BusinessProgress is a read projection of source documents and posted facts.
// It never stores or accepts a separately editable completion percentage.
type BusinessProgressAccess struct {
	Sales      bool `json:"sales"`
	Production bool `json:"production"`
	Tasks      bool `json:"tasks"`
	WIP        bool `json:"wip"`
}

type BusinessProgressQuery struct {
	View, Keyword, Scope, Risk, Owner, DateFrom, DateTo string
	Limit, Offset, ID                                   int
	SnapshotAt                                          time.Time
	Access                                              BusinessProgressAccess
	TaskVisibility                                      *WorkflowTaskVisibilityScope
}

type BusinessProgressCounts struct {
	Total   int `json:"total"`
	Overdue int `json:"overdue"`
	DueSoon int `json:"due_soon"`
	Blocked int `json:"blocked"`
	Undated int `json:"undated"`
}

type BusinessProgressRow struct {
	ID                int     `json:"id"`
	View              string  `json:"view"`
	OrderNo           string  `json:"order_no"`
	Status            string  `json:"status"`
	Customer          string  `json:"customer"`
	Product           string  `json:"product"`
	ProductCount      int     `json:"product_count"`
	ProductID         int     `json:"product_id"`
	SalesOwner        string  `json:"sales_owner"`
	DueDate           string  `json:"due_date"`
	UpdatedAt         string  `json:"updated_at"`
	Active            bool    `json:"active"`
	Overdue           bool    `json:"overdue"`
	DueSoon           bool    `json:"due_soon"`
	Blocked           bool    `json:"blocked"`
	Unlinked          bool    `json:"unlinked"`
	Unit              string  `json:"unit"`
	OrderedQuantity   *string `json:"ordered_quantity"`
	ShippedQuantity   *string `json:"shipped_quantity"`
	CompletedQuantity *string `json:"completed_quantity"`
	CurrentOperation  string  `json:"current_operation"`
	OperationCount    int     `json:"operation_count"`
	DeliveryKnown     bool    `json:"delivery_known"`
	EngineeringTotal  int     `json:"engineering_total"`
	EngineeringReady  int     `json:"engineering_ready"`
	ProductionOrders  int     `json:"production_orders"`
	ProductionClosed  int     `json:"production_closed"`
	MaterialTotal     int     `json:"material_total"`
	MaterialPending   int     `json:"material_pending"`
	InProgressBatches int     `json:"in_progress_batches"`
	OutsourcedBatches int     `json:"outsourced_batches"`
	WaitingBatches    int     `json:"waiting_batches"`
	RejectedBatches   int     `json:"rejected_batches"`
	PlannedBatches    int     `json:"planned_batches"`
	OpenTasks         int     `json:"open_tasks"`
	BlockedTasks      int     `json:"blocked_tasks"`
	UnassignedTasks   int     `json:"unassigned_tasks"`
	AttentionTaskID   int     `json:"attention_task_id"`
	AttentionTask     string  `json:"attention_task"`
	AttentionReason   string  `json:"attention_reason"`
	AttentionOwner    string  `json:"attention_owner"`
	AttentionRole     string  `json:"attention_role"`
}

type BusinessProgressBoard struct {
	Rows       []*BusinessProgressRow `json:"rows"`
	Total      int                    `json:"total"`
	Counts     BusinessProgressCounts `json:"counts"`
	SnapshotAt time.Time              `json:"snapshot_at"`
	Access     BusinessProgressAccess `json:"access"`
}

// Detail records contain navigation context only; editable documents remain on
// their domain pages. Sections are independently bounded and report truncation.
type BusinessProgressRecord struct {
	ID       int    `json:"id"`
	Kind     string `json:"kind"`
	Number   string `json:"number"`
	Label    string `json:"label"`
	Status   string `json:"status"`
	Quantity string `json:"quantity"`
	Unit     string `json:"unit"`
	Date     string `json:"date"`
	Owner    string `json:"owner"`
	Role     string `json:"role"`
	Note     string `json:"note"`
	ParentID int    `json:"parent_id"`
}

type BusinessProgressDetail struct {
	Row        *BusinessProgressRow                `json:"row"`
	Sections   map[string][]BusinessProgressRecord `json:"sections"`
	HasMore    map[string]bool                     `json:"has_more"`
	SnapshotAt time.Time                           `json:"snapshot_at"`
	Access     BusinessProgressAccess              `json:"access"`
}

type BusinessProgressRepo interface {
	ListBusinessProgress(context.Context, BusinessProgressQuery) (*BusinessProgressBoard, error)
	GetBusinessProgress(context.Context, BusinessProgressQuery) (*BusinessProgressDetail, error)
}

type BusinessProgressUsecase struct{ repo BusinessProgressRepo }

func NewBusinessProgressUsecase(repo BusinessProgressRepo) *BusinessProgressUsecase {
	return &BusinessProgressUsecase{repo: repo}
}

func normalizeBusinessProgressQuery(q BusinessProgressQuery) (BusinessProgressQuery, error) {
	q.Keyword, q.Owner = strings.TrimSpace(q.Keyword), strings.TrimSpace(q.Owner)
	if q.View == "" {
		q.View = "orders"
	}
	if q.Scope == "" {
		q.Scope = "active"
	}
	if q.Risk == "" {
		q.Risk = "all"
	}
	if q.Limit == 0 {
		q.Limit = 20
	}
	if q.View != "orders" && q.View != "production" ||
		q.Scope != "active" && q.Scope != "all" && q.Scope != "ended" ||
		q.Risk != "all" && q.Risk != "overdue" && q.Risk != "due_soon" && q.Risk != "blocked" && q.Risk != "undated" && q.Risk != "unlinked" ||
		q.Limit < 1 || q.Limit > 100 || q.Offset < 0 || q.Offset > 1000000 || q.ID < 0 ||
		len([]rune(q.Keyword)) > 100 || len([]rune(q.Owner)) > 100 {
		return q, ErrBadParam
	}
	if q.View == "orders" && !q.Access.Sales || q.View == "production" && !q.Access.Production {
		return q, ErrForbidden
	}
	if q.Access.Tasks && q.TaskVisibility == nil {
		return q, ErrForbidden
	}
	for _, date := range []string{q.DateFrom, q.DateTo} {
		if date != "" {
			if _, err := time.Parse("2006-01-02", date); err != nil {
				return q, ErrBadParam
			}
		}
	}
	if q.DateFrom != "" && q.DateTo != "" && q.DateFrom > q.DateTo {
		return q, ErrBadParam
	}
	if q.SnapshotAt.IsZero() {
		q.SnapshotAt = time.Now().UTC()
	}
	return q, nil
}

func (uc *BusinessProgressUsecase) List(ctx context.Context, q BusinessProgressQuery) (*BusinessProgressBoard, error) {
	q, err := normalizeBusinessProgressQuery(q)
	if err != nil {
		return nil, err
	}
	if uc == nil || uc.repo == nil {
		return nil, ErrBadParam
	}
	return uc.repo.ListBusinessProgress(ctx, q)
}

func (uc *BusinessProgressUsecase) Detail(ctx context.Context, q BusinessProgressQuery) (*BusinessProgressDetail, error) {
	q.Scope, q.Risk, q.Keyword, q.Owner, q.DateFrom, q.DateTo, q.Offset, q.Limit = "all", "all", "", "", "", "", 0, 1
	q, err := normalizeBusinessProgressQuery(q)
	if err != nil {
		return nil, err
	}
	if q.ID <= 0 || uc == nil || uc.repo == nil {
		return nil, ErrBadParam
	}
	return uc.repo.GetBusinessProgress(ctx, q)
}
